--- lisp/server.el	2013-02-13 07:17:58.000000000 +0300
+++ lisp/server.el	2014-03-01 11:56:56.818176609 +0300
@@ -1208,12 +1208,26 @@
 		  (cond
 		   ((and use-current-frame
 			 (or (eq use-current-frame 'always)
+                             ;; ORIGINAL CODE
+
 			     ;; We can't use the Emacs daemon's
 			     ;; terminal frame.
-			     (not (and (daemonp)
-				       (null (cdr (frame-list)))
-				       (eq (selected-frame)
-					   terminal-frame)))))
+                             ;; (not (and (daemonp)
+                             ;;           (null (cdr (frame-list)))
+                             ;;           (eq (selected-frame)
+                             ;;               terminal-frame)))))
+
+                             ;; If there is no X frames, set tty-name to
+                             ;; 'window-system and end with nil
+                             ;; in order to jump to the next clause of this cond
+                             (if (find-if (lambda (f)
+                                            (eq (framep f) 'x))
+                                          (frame-list))
+                                 t
+                               (and display
+                                    (progn 
+                                      (setq tty-name 'window-system)
+                                      nil)))))
 		    (setq tty-name nil tty-type nil)
 		    (if display (server-select-display display)))
 		   ((eq tty-name 'window-system)
